Atomic for BsModule

jonesmz opened the pull request.

jonesmz committed:
Atomic for BsModule

BearishSun commented:
Earlier on I decided against atomicity for Module. Primarily because it is something you end up paying for on every single call to Module::instance(), and it only really matters during startUp/shutDown calls, which happen once at very specific points.  

The memory barriers will prevent the compiler from optimally organizing the code around the relevant call, and can potentially issue a memory barrier instruction (will happen on ARM I believe, but on x86 just guaranteeing the order is enough). Perhaps if the atomicity could be achieved using relaxed memory ordering it would be acceptable but I'm not sure I'd like it even then though. 

If we were to go a thread safe route I'd prefer we use a separate version of Module that is thread-safe (perhaps via a policy controlled by a template parameter).

jonesmz commented:
Ok, that sounds good.

I'll add a set of bs::StoragePolicy classes that can be given as a template parameter.

Normal, which would just be the normal type being stored.
Thread local, which would be for thread local storage
Atomic, which would wrap the type in std::atomic.

hhyyrylainen committed:
Added externalWindowHandle support for Linux

This is pretty basic. This also most likely breaks input on external
windows as I left a bunch of the X related setup out because I don't
want bsf to do anything extra to my window. I use SDL2 for input. A
proper solution to this is probably having one more option for
specifying if bsf should run the window event loop or not, Windows
version probably also needs this option, but I haven't tested my code
on Windows yet.

BearishSun committed:
Bugfix: Don't try to retrieve the animation bone for the root scene object mapping, as it doesn't exist and causes a crash

BearishSun committed:
Bugfix: Deserialized renderables will not properly register their referenced resources for listening

BearishSun committed:
Bugfix: Fix binary comparer cleanup callback not triggering

BearishSun committed:
Bugfix: Reimporting a texture will now properly refresh its references in scene materials

BearishSun committed:
Docs: Updated CONTRIBUTORS.md

Commit:
Bugfix: Fix use of multiple descriptor sets in Vulkan in GpuParams

Commit:
WIP: MoltenVK
 - Fixed all shaders failing to cross compile to MSL
 - Added SPIR-V Cross and MoltenVK library dependencies
 - Added shader object remapping from descriptor set/slot to Metal indices
 - Perform vertex Y flip in shader to be consistent with Vulkan
 - Disable validation layers on macOS, as they complain about Metal Shading Language
 - Enable MoltenVK debug mode so XCode GPU debugging works

BearishSun committed:
Build: Update compiled shader dependencies package

Commit:
Bugfix: Fix invalid MSL remapping for non-structured buffers

Commit:
WIP: MoltenVK
 - Shader object remapping will now sort the objects based on set/slot, in order so they resulting indices match what MoltenVK expects

BearishSun committed:
WIP: MoltenVK
 - More fixes regarding set/slot rebinding
 - Explicitly set threadgroup size for MSL compute shaders

BearishSun committed:
Feature: Material variation can now be set from C#

BearishSun committed:
Feature: BSL now supports named variations
 - Such variations can be exposed to the editor and then toggled by the user directly from the GUI

BearishSun committed:
Feature: Added RTTIPlainType specialization for SmallVector

BearishSun committed:
Feature: A list of shader variation parameters and their allowed values can now be queried through Shader
 - Also added a 'show' BSL attribute that notifies the system a shader variation parameter should be exposed to the user (rather than controlled by the renderer)

BearishSun committed:
Bugfix: Fix invalid shader variation parameter meta-data serialization

BearishSun committed:
Bugfix: Reduced threadgroup size for indirect specular tiled deferred shader, as it was too large for macOS

BearishSun committed:
Bugfix: Cross-compiler updated so it emits less pessimistic memory barriers, significantly increasing performance on certain hardware under OpenGL/Vulkan

BearishSun committed:
Docs: Updated offscreen rendering manual to match the new Camera component API

BearishSun committed:
Bugfix: Fixed memory corruption due to invalid handling of arrays of structs in material params

BearishSun committed:
Feature: Custom search paths can now be added for shader includes

BearishSun committed:
Bugfix: Fix C# library compilation

BearishSun committed:
Bugfix: Fix a nasty threading issue that made command queue block operation not to block

BearishSun committed:
Bugfix: Fix animation curve splitting not working properly with a single-keyframe curves

BearishSun committed:
Bugfix: Fix BSL variation name & show attribute parsing

BearishSun committed:
Refactor: Added variation names to the Decal shader

BearishSun committed:
Bugfix: Fix script binding generation for struct interop objects containing arrays of other struct interop objects

jayrulez committed:
Feature: Exposed SceneObject hasFlag method to C#

BearishSun committed:
Bugfix: Independant blend for multiple render targets is now parsed properly in BSL

BearishSun committed:
Bugfix: Decal shader blend modes are now properly applied
 - Also hide the gMaskTex as internal

BearishSun committed:
Bugfix: Fix a crash when changing Decal's material

BearishSun committed:
Bugfix: Fix a crash when attempting to use a non-decal material with the Decal component

BearishSun committed:
Bugfix: Flip decal Y axis so it matches the texture along its local XY axes

BearishSun committed:
Feature: Vulkan backend will now report underlying element types for buffers & textures
 - Code for retrieving dummy buffer/image views can now use those values to create appropriate views

BearishSun committed:
Docs: Updated bullet point indentation

BearishSun committed:
Bugfix: GPU sort unit test now runs properly on Vulkan

BearishSun committed:
Feature: Added support for subgroup/warp/simd execution and memory barriers to BSL
 - Also added a [hideInInspector] attribute for hiding a BSL parameter from editor inspector GUI
 - Also add a [name("Nice name")] parameter for setting a pretty name for BSL parameters
 - Also updated Decal shader so it doesn't exibit artifacts on Metal due to use of 'discard'

BearishSun committed:
Bugfix: Fix BSL compiler not reporting an error if an include file cannot be found

BearishSun committed:
WIP: MoltenVK
 - Updated MoltenVK version
 - Don't use shared memory for buffers intended for reads, as its not supported by Metal
 - Don't use Vulkan 1.1 as it's not supported by MoltenVK

BearishSun committed:
Docs: Some doc tweaks

BearishSun committed:
Bugfix: Binding null (dummy) textures & buffers in Vulkan will now correctly register them with the command buffer and therefore issue layout transitions and relevant memory barriers

BearishSun committed:
Bugfix: Fix resource listener callbacks triggering on non-main thread
 - Also fix resource dependency queries not being thread safe

BearishSun committed:
Bugfix: Asset import tool can now process editor built-in assets properly

BearishSun committed:
Bugfix: Fix macOS release mode build

BearishSun committed:
Bugfix: Fix build on macOS

BearishSun committed:
Bugfix: Fix a deadlock when many resources are being loaded and all end up waiting for other resources, using up all scheduler threads

BearishSun committed:
Bugfix: Mesh and texture core thread readbacks should not trigger submits for other threads

BearishSun committed:
Bugfix: Decal size no longer returns size * scale instead of just local size
 - Also limit size to non-negative values

BearishSun committed:
Bugfix: Fix a crash when attempting to use a non-native texture format on Vulkan

BearishSun committed:
Bugfix: Fix sprite texture not being properly initialized

BearishSun committed:
Feature: Added better logging functionality BS_LOG that supports categories, more verbosity options and string formatting support

pgruenbacher-TSUS committed:
fix resource handle error default template argument

BearishSun committed:
Refactor: Updated a variety of logging calls to new BS_LOG API

BearishSun committed:
Refactor: Updated a variety of logging calls to new BS_LOG API

BearishSun committed:
Build: Fix gcc/clang compilation

Commit:
Build: Another fix for gcc/clang compilation

Commit:
Build: Fix OpenGL plugin compilation error

BearishSun committed:
Refactor: Cleanup and refactor some core thread functionality

BearishSun committed:
Feature: Core thread can now be optionally ran on the main thread
 - This ensures macOS renderer <-> windowing interactions work as intended due to their main thread requirements
 - Also the entry point for the framework as been updated and is now unified across all platforms

BearishSun committed:
Refactor: Updated C# debug code to match the new logging system

BearishSun committed:
Bugfix: Fix single-threaded rendering path

BearishSun committed:
Bugfix: Fix a threading issue that was causing secondary windows to render as blank on macOS

BearishSun committed:
Build: Fix macOS compilation issues

BearishSun committed:
Bugfix: Fix a potentially unsafe cross-thread operation on macOS

BearishSun committed:
Bugfix: Fix memory corruption on GUIElement destruction, due to incorrectly set update parents
 - Update parents were not letting their children know they were getting destroyed, which made the children call them when themselves were getting destroyed, causing an access after free.

BearishSun committed:
Bugfix: Fix helper GPU frames not freeing GPU profiler samples

BearishSun committed:
Feature: Blit shader can now be used for downsizing, instead of just copying and resolving samples

BearishSun committed:
Bugfix: When reading texture data on directx backend, respect the returned row and slice pitch
 - This issue caused read-back on textures with compressed formats to return incorrect pixel data

BearishSun committed:
Refactor: PixelData now stores row & slice pitch in bytes, rather than pixels
 - Original design was poor, as storing the pitch in pixels creates issues in how to represent it for block-compressed formats, which work in blocks rather than pixels

BearishSun committed:
Bugfix: Fix pixel row/slice pitch issues on Vulkan

BearishSun committed:
Bugfix: Fix read-back from render textures on Vulkan failing due to invalid layout
 - Optimal texture layout was assumed to be COLOR_ATTACHMENT, while the render pass was transitioning to SHADER_READ_ONLY instead

BearishSun committed:
Bugfix: Fix a crash when creating a 0-sized texture on OpenGL

BearishSun committed:
Feature: OpenGL backend now supports setting blend & color mask options on a per-target basis

BearishSun committed:
Refactor: All logging calls replaced with the new BS_LOG macro

BearishSun committed:
Build: Fixed some log calls missed by the previous commit

Commit:
Build: Fix Linux build broken by recent log changes

BearishSun committed:
Build: Fix macOS build issues caused by recent commit

BearishSun committed:
Bugfix: Fix NDC to Screen coordinate conversion for render backends with flipped NDC Y axis (i.e. Vulkan)

BearishSun committed:
Feature: Screen space lens flare
 - Also refactored scene downsampling code so its in its re-usable compositor node
 - Added a simple Composite shader for merging images
 - Fixed an issue with bloom using too large blur filter sizes

BearishSun committed:
Feature: C# API for the lens flare effect
 - Also fixed a compilation error with standalone script runner executable

BearishSun committed:
Feature: Bicubic upsampling shader

BearishSun committed:
Feature: User can now switch between bilinear and bicubic upsampling for lens flare

BearishSun committed:
Build: Fix Clang build broken with the lens flare feature

BearishSun committed:
Docs: Updated CONTRIBUTORS.md

BearishSun committed:
Bugfix: Avoid constantly rebuilding pipeline states on Vulkan
 - This was caused by underlying framebuffers being rebuilt (which is a different issue)
 - Render pass handling was split from framebuffer, so changing framebuffers will no longer trigger a pipeline state rebuild, as long as it has a matching render pass

BearishSun committed:
Refactor: Vulkan framebuffers no longer need different variations depending on clear/read/load flags
 - Instead we're now relying on render-pass compatibility to handle those, which gets rid of unnecessary overhead

BearishSun committed:
Refactor: Much improved GPU resource pooling

BearishSun committed:
Bugfix: Apply OpenGL scissor rect when it is set after the pipeline state

guestnone committed:
Refactor: Log Improvements
- Added bs_convert_timet_to_string function for time to string conversion.
- Move all time-to-string functions in bs::Time to bs_convert_timet_to_string.
- Fix the time not being correctly written to the log output.
- Added the function to save the log as a text file.
- Modify the Debug::saveLog function to allow for setting the type of the log.
- Added a class to manage and get the string names of the categories.
- Register all found categories during the framework initialization.
- Log verbosity now has a toString function.
- Fix the documentation to match the logging system revamp.
- Fixed some other minor things.

guestnone committed:
Refactor: Rename the time_t to String function to toString
- Move it to BsString and fixed code that used it.

guestnone committed:
Refactor: Merge LogCategoryName into Log class
- Port the code that used the old class to use a instance of the Log class in Debug.
- Fix the documentation to reflect the changes.

guestnone committed:
Other: Add default values in toString for time_t

BearishSun committed:
Refactor: Tweaks how log category names are registered

BearishSun committed:
Bugfix: Fix a crash caused by releasing COM objects twice, or attempting to release null pointers

BearishSun committed:
Bugfix: Fix material not refreshing after one of the textures it's referencing finishes loading

BearishSun committed:
Bugfix: Don't complain if width/height aren't a multiple of compressed block size, as this is okay in some cases (e.g. mip-maps smaller than block size)

pgruenbacher-TSUS committed:
imgui plugin cmake

pgruenbacher-TSUS committed:
fix cmake

pgruenbacher committed:
Update BsImgui.cmake

pgruenbacher-TSUS committed:
test travis with plugin

pgruenbacher-TSUS committed:
fix advanced variables

pgruenbacher-TSUS committed:
cmake compatibility

pgruenbacher-TSUS committed:
fix advanced

pgruenbacher-TSUS committed:
set name of animation clip resource when importing

BearishSun committed:
Bugfix: Fix Vulkan issues with 0-sized textures

BearishSun committed:
Bugfix: Fix euler -> quaternion animation curve conversion now properly handles rotations over 180 degrees
 - It no longer forces them to use the shorest path, causing them to wrap around

BearishSun committed:
Bugfix: Fix GUI line rendering under Vulkan
 - It was flipped on the Y axis

hhyyrylainen committed:
Made CMake install the needed bslinc raw files instead of asset versions

BearishSun committed:
Bugfix: Fix invalid image layout for Vulkan render targets in certain cases

ataulien committed:
Refactor: Make BS_LOG-makro work outside bs namespace

Also wraps the makro in do { ... } while(0) to make the compiler
complain about a missing semicolon after a BS_LOG invokation.

ataulien committed:
Fix build

Also wraps the makro in do { ... } while(0) to make the compiler
complain about a missing semicolon after a BS_LOG invokation.

ataulien committed:
Refactor: Make array RTTI macros work outside bs namespace

BearishSun committed:
Build: Fixing some Linux and macOS compile issues introduced with the previous commit

pgruenbacher-TSUS committed:
fix whitespacing. exclude binary files

ataulien committed:
Refactor: Make BS_LOG_VERBOSITY configurable from outside

BearishSun committed:
Refactor: Allow GUI style name passed using const char*

jayrulez committed:
Do not focus new non-modal window if there is any modal window present

pgruenbacher-TSUS committed:
update via comments

Commit:
Added CMake options to build with RTTI and exceptions and not strip

debug info. Default values should have the same behaviour as before.

BearishSun committed:
WIP: Initial work on Bokeh depth of field

BearishSun committed:
Refactor: Standardized magnitude checks when normalizing vectors
 - Also added a missing check when normalizing a quaternion

BearishSun committed:
Bugfix: Euler -> quaternion curve conversion will now properly preserve multiple spins between two keyframes
 - i.e. keyframes that are 360 or more degrees apart

BearishSun committed:
Build: Fix build issue on GCC/Clang introduced with recent commits

BearishSun committed:
Bugfix: Fixing a memory corruption issue in one of the script helper classes

BearishSun committed:
Docs: Updated animation utility comment to better reflect the current state
 - We're no longer converting between euler & quaternion curves back and forth, resolving the concern expressed in that comment

BearishSun committed:
Build: Make sure BSF_ENABLE_EXCEPTIONS is respected on MSVC

BearishSun committed:
Bugfix: Fix an issue when generating diffs of script types containing enums or resource references

BearishSun committed:
Build: Move generated .csproj files into the build folder, so that multiple build configurations dont conflict with each other

BearishSun committed:
Bugfix: Fix swapped middle and right mouse buttons in the event button states

BearishSun committed:
Bugfix: Fix C# project generation on Windows

BearishSun committed:
WIP: Bokeh DOF
 - Getting bokeh shaders to compile and backing C++ code

BearishSun committed:
Build: Fix shader include install

BearishSun committed:
Bugfix: Fix a crash when attempting a normal write on a dynamic texture on DirectX

BearishSun committed:
WIP: Bokeh DOF
 - Added C++ code for the 'combine' shader
 - Fixed some compile issues with the 'combine' shader
 - Added a renderer node combining all the shaders

BearishSun committed:
Refactor: Gamma <-> linear conversion moved to Color type

BearishSun committed:
Feature: Added a [HDR] attribute to mark colors as supporting high dynamic range

BearishSun committed:
Refactor: Added a 'create' method to ShaderImportOptions

BearishSun committed:
Refactor: Added HDR color variants of ColorGradient and ColorDistribution
 - Updated Material to use the HDR variant by default

BearishSun committed:
Refactor: Whitespace changes on script generated code

BearishSun committed:
Refactor: Added ColorGradientHDR serializable type

BearishSun committed:
WIP: Bokeh DOF
 - Added default Bokeh flare texture to the renderer

BearishSun committed:
Bugfix: Don't crash if renderer material cannot be loaded

BearishSun committed:
Bugfix: When doing mesh/texture readback make sure other core thread operations are submitted
 - In particular it needs to guarantee that any initialize() calls originating from other threads get processed
 - This might break some functionality as this was previously change to fix some other issue

BearishSun committed:
Build: Fix Linux build

BearishSun committed:
Bugfix: Fix BSL compiler not stripping [hdr] attribute from HLSL

nemerle committed:
Feature: extended `GpuResourceData` interface with `setData` function

Refactor: introduce `verifyLockAndThread` helper function to reduce repetition.

This should make modifying contents easier and reduce the need to allocate
fully blown PixelData to implement simple operations ( flipping images etc. )

hhyyrylainen committed:
Fixed MSVC crash and error about structured exceptions with exceptions
enabled with visual studio

BearishSun committed:
WIP: Bokeh DOF
 - Many various fixes, tweaks and additions in order to get the effect to work

BearishSun committed:
WIP: Networking
 - A nice chunk of high-level code for the replication system prototype

BearishSun committed:
Bugfix: Fixed an issue that caused no rendering to output unless Bokeh DOF was turned on

BearishSun committed:
Bugfix: Fix Gaussian DOF not rendering, broken with Bokeh DOF refactor

BearishSun committed:
WIP: Bokeh DOF
 - Fix rendering on OpenGL
 - Fix slightly incorrect normalization weights
 - Improve layer blending to reduce bleeding and abrupt changes
 - Fixed incorrect bokeh flare texture URL

BearishSun committed:
Bugfix: Fix cross-compiler generating code that doesn't compile in Vulkan, due to invalid float literal generation

BearishSun committed:
WIP: Bokeh DOF
 - Added MSAA support

BearishSun committed:
Refactor: Minor tweak to resolve a warning in VS2019

BearishSun committed:
Bugfix: Fix C# code broken due to multiple RenderSettings ctors being exported

BearishSun committed:
Bugfix: Fix broken SerializedObject creation, always assuming 'shallow' serialization mode

BearishSun committed:
Bugfix: Fix an issue with character input stopping to work after ctrl or alt keys get stuck on Windows
 - Especially after alt+tab

BearishSun committed:
Bugfix: Fix a crash during Material deserialization

BearishSun committed:
Feature: Bokeh depth-of-field effect
 - Final bit of work on Bokeh DOF
 - Added soft depth occlusion in order to reduce bright background bleedthrough

BearishSun committed:
Refactor: Added [hdr] attribute to all emissive color parameters in shaders

BearishSun committed:
Refactor: Updated C# code to report the HDR attribute on shader parameters

BearishSun committed:
Refactor: MemoryDataStream can now auto-grow
 - Preparing data streams to be used in all RTTI types, and having BitStream implement the DataStream interface

BearishSun committed:
Refactor: Degree and Radian headers are no longer unnecessarily included in main prerequisites header

BearishSun committed:
Refactor: Removing RTTIPlainType specializations from non-RTTI headers

BearishSun committed:
Refactor: More RTTI code moved out of the non-RTTI headers

BearishSun committed:
Refactor: Renamed Prerequisites/BsRTTIPrerequisites.h to Reflection/BsRTTIPlain.h
 - It better represents the current usage of that file, as it now contains plain RTTI related code

BearishSun committed:
Refactor: Moving RTTIPlainType to use Bitstream as source/destination, instead of raw memory

BearishSun committed:
Refactor: Updated more RTTI types to the new RTTI plain type structure

Commit:
Build: Fix for gcc/clang compilation broken with previous commits

hhyyrylainen committed:
Added a basic application alternative that allows rendering one frame

at a time

hhyyrylainen committed:
Moved the single frame stuff to CoreApplication

BearishSun committed:
WIP: Motion blur
 - Added basic camera movement based motion blur

BearishSun committed:
Refactor: Moving RTTIPlainType to use Bitstream as source/destination, instead of raw memory

BearishSun committed:
Bugfix: Fixing string plain RTTI broken with recent changes

BearishSun committed:
WIP: Motion blur
 - Updated shader binaries with motion blur shader

BearishSun committed:
Bugfix: Fixed animation looping if not all curves are the same length

BearishSun committed:
Bugfix: Fix Animation::getState returning invalid layer index

BearishSun committed:
Feature: When a window loses focus remember focused GUI elements and restore them when the window regains focus

BearishSun committed:
WIP: Motion blur
 - Added missing PPMotionBlur shader

BearishSun committed:
Bugfix: Fixing camera view sometimes getting stuck after render settings change

BearishSun committed:
Bugfix: Fix for GPU parameters being parsed/set incorrectly if a color parameter used less than 4 channels

BearishSun committed:
Bugfix: Do not enable motion blur by default

BearishSun committed:
Bugfix: Fixed a crash if a skybox attempts to be initialized without the referenced cubemap texture being fully loaded

BearishSun committed:
Bugfix: Fixed a crash caused by FBX importer when the imported mesh has no material indices defined

BearishSun committed:
Bugfix: Fixed a crash when setting screen space lens flare's 'downsampleCount' setting to zero

BearishSun committed:
Bugfix: Make sure WIP motion blur rendering is disabled

hhyyrylainen committed:
Added callbacks for crash handling and log messages

BearishSun committed:
WIP: Initial work on GUI draw results caching

BearishSun committed:
Refactor: Renaming rttiWrite/Read/GetSize methods to match the naming scheme of other globals

BearishSun committed:
Docs: Various doc cleanup and updates

BearishSun committed:
Docs: Fixed an out of date doc comment

BearishSun committed:
Build: Fixing mac build

hhyyrylainen committed:
Added return values for the early SEH exits

hhyyrylainen committed:
Changed crash reporter settings to be stored as struct in the CrashHandler

hhyyrylainen committed:
Moved the custom log callback to Debug

BearishSun committed:
Refactor: Added a 'compress' flag to all plain RTTI types

hhyyrylainen committed:
Removed the extra frame render before shutdown

If anyone else gets crashes "when doing it correctly" then they need
to add one extra runMainLoopFrame call before calling endMainLoop

hhyyrylainen committed:
Removed copydoc from Debug::mCustomLogCallback

BearishSun committed:
Refactor: Refactoring binary serializer to use streams

BearishSun committed:
WIP: GUI caching
 - Moved most of the caching code to separate GUIDrawGroups class
 - Work on moving mesh updates to GUIDrawGroups and making them per-widget

BearishSun committed:
WIP: GUI caching
 - Worked on core thread draw group updates
 - Building/rebuilding draw group texture

BearishSun committed:
Refactor: GUI render elements refactored with a cleaner approach with less code and virtual calls

BearishSun committed:
Bugfix: Allow crash handler to be created with no settings object, as this is preventing some existing executables from compiling

BearishSun committed:
WIP: GUI caching
 - Updated the code to work with the GUI render element refactor code

BearishSun committed:
WIP: GUI caching
 - Work on getting GUI element changes to get recognized by the draw group system

BearishSun committed:
WIP: GUI caching
 - Added code for rebuilding and dirtying GUI draw groups when GUI element is reported as dirty

hhyyrylainen committed:
Replaced type punning with a memcpy

Silences a warning from GCC. AFAIK this is the only correct way to
read the data through a different type "pointer."

BearishSun committed:
WIP: GUI caching
 - Some cleanup

BearishSun committed:
Bugfix: Fix memory corruption when serializing shaders, broken with recent RTTI refactor

hhyyrylainen committed:
Added methods to render window to notify it about external events,

needed when using an external window handle.

BearishSun committed:
WIP: GUI caching
 - Generate and send dirty GUI render data to the GUI renderer

BearishSun committed:
WIP: GUI caching
 - Updated mesh generation and rendering to handle the new draw group system
 - Also updated GUI docs to match recent changes

BearishSun committed:
WIP: Networking
 - Laying the groundwork for binary serializer refactor that supports compression

BearishSun committed:
WIP: Networking
 - Added support for RTTI schema
 - Cleaned up RTTIField interface a bit

BearishSun committed:
WIP: Networking
 - RTTIPlainType::getDynamicSize -> getSize

BearishSun committed:
Feature: GUI caching
 - Fixed compositing of draw groups and transparency handling
 - Extended RendererExtension so it is allowed to change and restore the renderer render target
 - Various fixes in order to get caching rendering properly

BearishSun committed:
Feature: RenderTarget now has an update count that increments any time something is drawn to it

BearishSun committed:
Feature: Camera on-demand drawing
 - RenderExtension was also updated so it can avoid rendering if it has no relevant data
 - Renderer has been updated so it can avoid swap completely if nothing has been rendered that frame
 - Also updated core sync methods to match the equivalent rtti method naming

BearishSun committed:
Bugfix: Fix clipping not properly clipping x/y coordinates in certain cases

BearishSun committed:
Build: Removing Imgui plugin from CI

BearishSun committed:
Build: Removing Imgui plugin from CI

BearishSun committed:
Bugfix: Ensure GUI widgets are sorted by depth when drawing using the new approach
 - Also make sure all widgets redraw when requested by the extension

BearishSun committed:
WIP: Networking
 - Buffered bitstream now supports align() and var int
 - Included a header for RTTI schema I missed previously

BearishSun committed:
WIP: Networking
 - Refactored binary serializer so it can read meta-data from a schema

BearishSun committed:
WIP: Networking
 - RTTI plain type sizes are now represented through BitLength class
 - RTTI plain type header can now be compressed

BearishSun committed:
WIP: Networking
 - rtti_* methods now return BitLength

BearishSun committed:
WIP: Networking
 - BinarySerializer can now handle compressed reflectable and reflectable pointer type sizes

BearishSun committed:
WIP: Networking
 - Extended RTTIPlainType::getSize with RTTIFieldInfo parameter
 - Fixed all compile errors

BearishSun committed:
Bugfix: Fixed an issue that caused GUI elements not to redraw after layout updates

BearishSun committed:
Bugfix: Fixed a crash caused by invalid GUI widget depth sorting

BearishSun committed:
Bugfix: Fix renderer extensions not properly redrawing

BearishSun committed:
Bugfix: Make sure GUI elements displaying a RenderTexture will properly refresh when it updates

BearishSun committed:
Build: Fix a build error on non-MSVC

BearishSun committed:
Build: Regenerated script bindings with new version of the script generator tool

Fishrock123 committed:
Meta: add DCO-1.1 to CONTRIBUTING.md

The DCO is a legal protection for some licenses (such as MIT) which do not expressly verify contributors' consent.

MIT also does not have a patents clause, which the DCO helps address.

This is not retroactive, but could help protect future contributions.

BearishSun committed:
Bugfix: Fixed an issue that would cause draw queue to keep piling up with renderables when an on-demand camera was not drawing

BearishSun committed:
Bugfix: Don't try to render GUI render elements with 0 vertices

BearishSun committed:
WIP: Networking
 - Added support for builtin plain types to BinarySerializer

BearishSun committed:
Docs: Updated manuals with new RTTI changes

BearishSun committed:
Bugfix: Fix invalid plain type serialization for integer types

BearishSun committed:
Bugfix: Fix recently broken DataBlock deserialization

BearishSun committed:
Refactor: Camera's OnDemand flag shouldn't be serializer

BearishSun committed:
Bugfix: On-demand drawing can now persist for multiple frames to accomodate multi-frame effects such as auto-exposure or temporal AA

BearishSun committed:
Bugfix: Fix GUI rendering on OpenGL after recent refactor
 - Also GUI draw group bounds are now properly calculated, not wasting space

BearishSun committed:
Feature: Added completion states on command buffers, so callers know when they finished executing on the GPU
 - Also updated the GPU profiler so individual views can be profiled independently (profiling entire frames was causing issues with queries across multiple command buffers)

BearishSun committed:
Refactor: DirectX and OpenGL render backends now use command buffers internally, so the command execution can be tracked

BearishSun committed:
Feature: Track auto-exposure progress on the CPU so we can determine how long to keep redrawing an on-demand view for
